### PR TITLE
feat(levm): vm interface

### LINKED ISSUE #578: LEVM: Move vm_tests and operations_tests to tests

### LINKED ISSUE #607: LEVM: Modify VM public interface
The LEVM should be created with a Tx type that already includes the `sender`.

### PR DESCRIPTION
**Pull request by lazcanoluca**
**Motivation**

To scaffold the necessary environment to execute a transaction provided a transaction environment, a block environment and a world state.

**Description**

- Add `TxEnv` struct to support the transaction environment.
- Change the EVM API to be created from a `TxEnv`, a `BlockEnv` and a `Db`.

Closes #607 
Closes #578

### CODE REVIEW

File: crates/levm/src/transaction.rs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    //
+    // [EIP-155]: https://eips.ethereum.org/EIPS/eip-155
+    // pub chain_id: Option<u64>,
+
+    // A list of addresses and storage keys that the transaction plans to access.
+    //
+    // Added in [EIP-2930].
+    //
+    // [EIP-2930]: https://eips.ethereum.org/EIPS/eip-2930
+    pub access_list: AccessList,
```
Comment (iinaki) [inline crates/levm/src/transaction.rs:47]: This should be wrapped in an Option, if the tx is of type Legacy there is no AccessList

File: crates/levm/src/transaction.rs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    Call(Address),
+    /// Contract creation.
+    Create,
+}
+
+/// The transaction environment.
+#[derive(Clone, Debug)]
+pub struct TxEnv {
+    /// Caller aka Author aka transaction signer.
+    pub caller: Address,
```
Comment (iinaki) [inline crates/levm/src/transaction.rs:20]:
```suggestion
    pub msg_sender: Address,
```
Let's rename this to msg_sender to keep the same naming everywhere

File: crates/levm/src/transaction.rs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    /// The gas limit of the transaction.
+    pub gas_limit: u64,
+    /// The gas price of the transaction.
+    pub gas_price: U256,
```
Comment (iinaki) [inline crates/levm/src/transaction.rs:24]: gas_price should be an Option as well, as it only exists on the types Legacy and AccessList

File: crates/levm/src/transaction.rs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    // The nonce of the transaction.
+    pub nonce: Option<u64>,
+    // Caution: If set to `None`, then nonce validation against the account's nonce is skipped: [InvalidTransaction::NonceTooHigh] and [InvalidTransaction::NonceTooLow]
+
+    // The chain ID of the transaction. If set to `None`, no checks are performed.
+    //
+    // Incorporated as part of the Spurious Dragon upgrade via [EIP-155].
+    //
+    // [EIP-155]: https://eips.ethereum.org/EIPS/eip-155
+    // pub chain_id: Option<u64>,
```
Comment (iinaki) [inline crates/levm/src/transaction.rs:40]: Here we left the chain_id commented

File: crates/levm/src/transaction.rs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    // [EIP-2930]: https://eips.ethereum.org/EIPS/eip-2930
+    pub access_list: AccessList,
+
+    /// The priority fee per gas.
+    ///
+    /// Incorporated as part of the London upgrade via [EIP-1559].
+    ///
+    /// [EIP-1559]: https://eips.ethereum.org/EIPS/eip-1559
+    /// aka `max_priority_fee_per_gas` or _miner tip_
+    pub gas_priority_fee: Option<U256>,
```
Comment (iinaki) [inline crates/levm/src/transaction.rs:55]: I would stick with the name `max_priority_fee_per_gas` as it is the same used in the EF tests

File: crates/levm/src/vm.rs [hunk 6]
```diff
@@ -115,6 +132,7 @@ pub struct VM {
     pub accrued_substate: Substate,
     /// Mapping between addresses (160-bit identifiers) and account
     /// states.
+    // pub state: WorldState,
```
Comment (iinaki) [inline crates/levm/src/vm.rs:135]: I see that we are keeping the `Db` over the `WorldState`, I think the choice is fine, but if so we should drop the `WorldState` abstraction so there is no confusion

File: crates/levm/src/vm.rs [hunk 7]
```diff
@@ -1295,6 +1307,10 @@ impl VM {
         self.call_frames.last().unwrap()
     }
 
+    pub fn add_account(&mut self, address: Address, account: Account) {
```
Comment (iinaki) [inline crates/levm/src/vm.rs:1310]: In `Db` we have an `add_account` so we can drop this method

File: crates/levm/src/vm.rs [hunk 8]
```diff
@@ -137,33 +155,27 @@ fn address_to_word(address: Address) -> U256 {
...
-        let initial_account = Account::new(balance, bytecode.clone());
+    pub fn new(tx_env: TxEnv, block_env: BlockEnv, db: Db) -> Self {
+        let bytecode = match tx_env.transact_to {
+            TransactTo::Call(addr) => db.get_account_bytecode(&addr),
+            TransactTo::Create => {
+                todo!()
+            }
+        };
 
         let initial_call_frame = CallFrame::new_from_bytecode(bytecode);
```
Comment (iinaki) [inline crates/levm/src/vm.rs:166]: Here I would use the data that we bring from the `TxEnv` to create a new `CallFrame` in the same way that we do on `generic_call`, with the method `CallFrame::new`.
```suggestion
        let initial_call_frame = CallFrame::new(
            gas,
            msg_sender,
            to,
            code_address,
            delegate,
            callee_bytecode,
            value,
            calldata,
            is_static,
        );
```

**Review by iinaki — COMMENTED** [review]
I left a few comments so as to unify this stuff and with the part from the EF tests, but in overall good work!

File: crates/levm/src/transaction.rs (hunk 1 above)
Comment (lazcanoluca) [reply crates/levm/src/transaction.rs:47]: True!

File: crates/levm/src/transaction.rs (hunk 2 above)
Comment (lazcanoluca) [reply crates/levm/src/transaction.rs:20]: Ok!

File: crates/levm/src/transaction.rs (hunk 3 above)
Comment (lazcanoluca) [reply crates/levm/src/transaction.rs:24]: It is nonetheless calculated at some point (effective gas price). I'll wrap it in a option for now, and then we will see how to adapt it.

File: crates/levm/src/transaction.rs (hunk 4 above)
Comment (lazcanoluca) [reply crates/levm/src/transaction.rs:40]: I'll uncomment!

File: crates/levm/src/transaction.rs (hunk 5 above)
Comment (lazcanoluca) [reply crates/levm/src/transaction.rs:55]: Good call!

File: crates/levm/src/vm.rs (hunk 6 above)
Comment (lazcanoluca) [reply crates/levm/src/vm.rs:135]: True

File: crates/levm/src/vm.rs (hunk 7 above)
Comment (lazcanoluca) [reply crates/levm/src/vm.rs:1310]: Oops, it was left from some merge. Fixed!

File: crates/levm/src/vm.rs (hunk 8 above)
Comment (lazcanoluca) [reply crates/levm/src/vm.rs:166]: I created it with some of the `tx_env` fields, but I left some as default for now.

**Review by iinaki — APPROVED** [review]
LGTM

**Review by jrchatruc — APPROVED** [review]

**Pull request merged**